File: src/main/resources/electronico/src/app/service/product.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Product} from "../model/product.model";
import {Observable} from "rxjs";

@Injectable({
  providedIn: 'root'
})
export class ProductService {

  constructor(private http: HttpClient) { }

  add(product: Product): Observable<any> {
    return this.http.post('http://localhost:8084/api/product/add', product);
  }

  get(productId: number): Observable<any> {
    return this.http.get('http://localhost:8084/api/product/get/' + productId);
  }

  update(product: Product): Observable<any> {
    return this.http.put('http://localhost:8084/api/product/update', product);
  }

  uploadImage(productId: number, imageData: FormData): Observable<any> {
    return this.http.post('http://localhost:8084/api/product/upload-image/' + productId, imageData);
  }

  getAll(): Observable<any> {
    return this.http.get('http://localhost:8084/api/product/get/all');
  }

  searchByTitle(title: string): Observable<any> {
    return this.http.get('http://localhost:8084/api/product/searchByTitle/' + title);
  }

  searchByCategory(category: string): Observable<any> {
    return this.http.get('http://localhost:8084/api/product/searchByCategory/' + category);
  }

  searchByManufacturer(manufacturer: string): Observable<any> {
    return this.http.get('http://localhost:8084/api/product/searchByManufacturer/' + manufacturer);
  }

  getRating(productId: number): Observable<any> {
    return this.http.get('http://localhost:8084/api/product/rating/' + productId);
  }
}

File: src/main/java/com/example/electronico/service/interfaces/ProductService.java
package com.example.electronico.service.interfaces;

import com.example.electronico.exception.ProductNotFoundException;
import com.example.electronico.exception.UserNotFoundException;
import com.example.electronico.model.Product;
import org.springframework.http.ResponseEntity;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;

public interface ProductService {
    Product add(Product product);

    Product get(Long productId) throws ProductNotFoundException;

    List<Product> getAll();

    Product update(Product product) throws ProductNotFoundException;

    ResponseEntity<String> delete(Long productId) throws ProductNotFoundException;

    Product uploadImage(MultipartFile file, Long productId) throws IOException, UserNotFoundException;

    List<Product> searchByTitle(String title);

    List<Product> searchByCategory(String category);

    List<Product> searchByManufacturer(String manufacturer);
}
